cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaQuorumCheck.java

```diff
@@ -0,0 +1,132 @@
+    private final Reconciliation reconciliation;
+    private final Admin admin;
+    private final Vertx vertx;
+    private final long controllerQuorumFetchTimeoutMs;
+
+    protected KafkaQuorumCheck(Reconciliation reconciliation, Admin ac, Vertx vertx, long controllerQuorumFetchTimeoutMs) {
+        this.reconciliation = reconciliation;
+        this.admin = ac;
+        this.vertx = vertx;
+        this.controllerQuorumFetchTimeoutMs = controllerQuorumFetchTimeoutMs;
+    }
+
+    /**
+     * Returns future that completes with true if the given controller can be rolled based on the quorum state. Quorum is considered
+     * healthy if the majority of controllers have caught up with the quorum leader within the controller.quorum.fetch.timeout.ms.
+     */
+    Future<Boolean> canRollController(int nodeId) {
+        LOGGER.debugCr(reconciliation, "Determining whether controller pod {} can be rolled", nodeId);
+        return describeMetadataQuorum().map(info -> {
+            boolean canRoll = isQuorumHealthyWithoutNode(nodeId, info);
+            if (!canRoll) {
+                LOGGER.debugCr(reconciliation, "Not restarting controller pod {}. Restart would affect the quorum health", nodeId);
+            }
+            return canRoll;
+        }).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining whether it is safe to restart controller pod {}", nodeId, error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    /**
+     * Returns id of the quorum leader.
+     **/
+    Future<Integer> quorumLeaderId() {
+        LOGGER.debugCr(reconciliation, "Determining the controller quorum leader id");
+        return describeMetadataQuorum().map(QuorumInfo::leaderId).recover(error -> {
+            LOGGER.warnCr(reconciliation, "Error determining the controller quorum leader id", error);
+            return Future.failedFuture(error);
+        });
+    }
+
+    /**
+     * Returns true if the majority of the controllers' lastCaughtUpTimestamps (not including the given node) are within
+     * the controller.quorum.fetch.timeout.ms based on the given quorum info.
+     **/
+    private boolean isQuorumHealthyWithoutNode(int nodeId, QuorumInfo info) {
+        int leaderId = info.leaderId();
+        if (leaderId < 0) {
+            LOGGER.warnCr(reconciliation, "No controller quorum leader is found because the leader id is set to {}", leaderId);
+            return false;
+        }
+
+        Map<Integer, Long> controllerStates = info.voters().stream().collect(Collectors.toMap(
+                QuorumInfo.ReplicaState::replicaId,
+                state -> state.lastCaughtUpTimestamp().isPresent() ? state.lastCaughtUpTimestamp().getAsLong() : -1));
+        int totalNumOfControllers = controllerStates.size();
+
+        if (totalNumOfControllers == 1) {
+            LOGGER.warnCr(reconciliation, "Performing rolling update on a controller quorum with a single node. The cluster may be " +
```

Comment:
How about a warn message like
```
Performing rolling update on a controller quorum with a single node. This may result in data loss " +
                    "or impact stability of the cluster during the rolling update. It is recommended that a minimum of three controllers are used.
```
I'm happy to put additional comments in the code as well.